src/core/internal/utf.d

```diff
@@ -425,61 +425,98 @@ dchar decode(in dchar[] s, ref size_t idx)
         return c; // dummy return
     }
 
-
 /* =================== Encode ======================= */
 
-/*******************************
- * Encodes character c and appends it to array s[].
+/**
+ * Encodes `c` into the static array `buf`.
+ *
+ * Params:
+ *     buf = destination of encoded character
+ *     c = character to encode
+ *
+ * Returns:
+ *     The length of the encoded character (a number between `1` and `4` for
+ *     `char[4]` buffers and a number between `1` and `2` for `wchar[2]` buffers)
+ *     or `0` in case of failure.
```

Comment:
Great idea. Let's define this function for all inputs (no contracts, no `assert(isValidDChar)`) and have it return 0 upon bad dchar. Then higher-level functions can throw exceptions, use replacement char etc.

Reply:
Additionally, instead of returning a length, return a slice to the buffer, or `null` on bad `dchar`, because the next thing people will need to do is slice that buffer.
Returning `null` and *not* an empty slice is important, because otherwise `if (auto s = encode(...))` will always be `true`.
Lastly, since this piece of code is highly performance sensitive, did you inspect the assembly to make sure `out` does not get default-initialized on function entry ?

Reply:
Isn't it better to pass `buf` as a regular `char[]` to push the allocation choice to the caller?